from Bio import SeqIO
import pandas as pd
import numpy as np
from tqdm import tqdm
from Bio import pairwise2

def local_align_score(seq1, seq2):
    alignments = pairwise2.align.localms(
        seq1, seq2,
        2,    # match
        -1,   # mismatch
        -5,   # gap open
        -0.5  # gap extend
    )
    return alignments[0].score if alignments else 0

def normalized_similarity(seq1, seq2):
    raw = local_align_score(seq1, seq2)
    return raw / min(len(seq1), len(seq2))

def match_sequences(ds_id, ds_seq, seqs_silkome, top_k=5):
    matches = {}

    scores = []

    for silk_id, silk_seq in tqdm(seqs_silkome.items(), desc=f'Aligning {ds_id}'):
        sim = normalized_similarity(ds_seq, silk_seq)
        #print("Similarity score: ", sim)
        scores.append((silk_id, sim))

    # Sort by similarity (descending)
    scores.sort(key=lambda x: x[1], reverse=True)

    # Keep top-k
    matches[ds_id] = scores[:top_k]
        
    return matches

def matches_to_dataframe(matches):
    rows = []

    for ds_id, hits in matches.items():
        for rank, (silk_id, score) in enumerate(hits, start=1):
            rows.append({
                "ds_id": ds_id,
                "rank": rank,
                "silkome_id": silk_id,
                "similarity": score
            })

    return pd.DataFrame(rows)


if __name__ == '__main__':
    seq_nt = 'MGHHHHHHSHTTPWTNPGLAENFMNSFMQGLSSMPGFTASQLDDMSTIAQSMVQSIQSLAAQGRTSPNKLQALNMAFASSMAEIAASEEGGGSLSTKTSSIASAMSNAFLQTTGVVNQPFINEITQLVSMFAQAGMNDVSA'
    seq_ct = 'VTSGGYGYGTSAAAGAGVAAGSYAGAVNRLSSAEAASRVSSNIAAIASGGASALPSVISNIYSGVVASGVSSNEALIQALLELLSALVHVLSSASIGNVSSVGVDSTLNVVQDSVGQYVG'
    seqs_ds = {}
    seqs_ds['NT2RepCT'] = 'MGHHHHHHMSHTTPWTNPGLAENFMNSFMQGLSSMPGFTASQLDDMSTIAQSMVQSIQSLAAQGRTSPNKLQALNMAFASSMAEIAASEEGGGSLSTKTSSIASAMSNAFLQTTGVVNQPFINEITQLVSMFAQAGMNDVSAGNSGRGQGGYGQGSGGNAAAAAAAAAAAAAAAGQGGQGGYGRQSQGAGSAAAAAAAAAAAAAAGSGQGGYGGQGQGGYGQSGSVTSGGYGYGTSAAAGAGVAAGSYAGAVNRLSSAEAASRVSSNIAAIASGGASALPSVISNIYSGVVASGVSSNEALIQALLELLSALVHVLSSASIGNVSSVGVDSTLNVVQDSVGQYVG'
    seqs_ds['A3IA'] = 'MGHHHHHHMSHTTPWTNPGLAENFMNSFMQGLSSMPGFTASQLDDMSTIAQSMVQSIQSLAAQGRTSPNKLQALNMAFASSMAEIAASEEGGGSLSTKTSSIASAMSNAFLQTTGVVNQPFINEITQLVSMFAQAGMNDVSAGNSGRGQGGYGQGSGGNAAAIAAAIAAIAAAAGQGGQGGYGRQSQGAGSAAAAAAAAAAAAAAGSGQGGYGGQGQGGYGQSGSVTSGGYGYGTSAAAGAGVAAGSYAGAVNRLSSAEAASRVSSNIAAIASGGASALPSVISNIYSGVVASGVSSNEALIQALLELLSALVHVLSSASIGNVSSVGVDSTLNVVQDSVGQYVG'
    seqs_ds['Rep1'] = seq_nt + 'GNSGPGPQGPSGPGPQGPYGPGPQGPGPQGPAPQGPSGPGPQRPQGPGPQRPYGPGGISVVSTTVSGPGPQGPSAPGPQGPYGPGPQVPGPQGPGPQGPSGPGPQRPQGPGPQGPYGPGGVSVVSQTVSGPGPQGPSGPGPQGPYGPGPQGPGPQGPGPQGPSGAGPQRPQGPGPQGPSGS' + seq_ct
    seqs_ds['Rep2'] = seq_nt + 'GNSGRPSSSYGAPGGGNGGRPSDTYGAPGGGNGGRPSDTYGAPGGGGNGNGGRPSSSYGAPGQGQGNGNGGRPSSSYGAPGGGNGGRPSDTYGAPGGGNGGRPSDTYGAPGGGNNGGRPSSSYGAPGGGNGGRPSDTYGAPGGGNSGS' + seq_ct
    seqs_ds['Rep3'] = seq_nt + 'GNSGGAGVPGVPGAIPGIGGIAGVGTPAAAAAAAAAAKAAKYGAAAGLVPGGPGFGPGVVGVPGAGVPGVGVPGAGIPVVPGAGIPGAAVPGVVSPEAAAKAAAKAAKYGARPGVGVGGSGS' + seq_ct
    seqs_ds['Rep5'] = seq_nt + 'GNSGRGQGGYGQGSGGNAAAAAAAAAAAAAAAGAGAAGVLPGVGGAGVPGVPGAIPGIGGIAGVGTPAAAAAAAAAAAAAAGAGAAGVLPGVGGAGVPGVPGAIPGIGGIAGVGTPSVTSGGYGYGTSAAAGAGVAAGSYAGAVNRLSSAEAASGS' + seq_ct
    seqs_ds['Rep7'] = seq_nt + 'GNSSSSTTTTTSAARSQAASQSASSSYSSAFAQAASSSFATSSALSRAFSSVSSASAASSLAYSIGLSAARSLGIADAAGLAGALARAVGALGQGATAASYGNALSTAAGQFFATAGLLNAGNASALASSFARAFSASAESQSFAQSQAFQQASAFQQAASRSASQSAAEADSTSSSGS' + seq_ct
    seqs_ds['4A 2rep'] = seq_nt + 'GRGQGGYGQGSGGNAAAAGQGGQGGYGRQSQGAGSAAAAGSGQGGYGGQGQGGYGQ' + seq_ct
    seqs_ds['VN-A3IA'] = 'MGHHHHHHMENLYFQGGPNSPQVTRGDVFTMPHMSHTTPWTNPGLAENFMNSFMQGLSSMPGFTASQLDDMSTIAQSMVQSIQSLAAQGRTSPNKLQALNMAFASSMAEIAASEEGGGSLSTKTSSIASAMSNAFLQTTGVVNQPFINEITQLVSMFAQAGMNDVSAGNSGRGQGGYGQGSGGNAAAIAAAIAAAIAAAGQGGQGGYGRQSQGAGSAAAAAAAAAAAAAAGSGQGGYGGQGQGGYGQSGSVTSGGYGYGTSAAAGAGVAAGSYAGAVNRLSSAEAASRVSSNIAAIASGGASALPSVISNIYSGVVASGVSSNEALIQALLELLSALVHVLSSASIGNVSSVGVDSTLNVVQDSVGQYVG'
    seqs_ds['fNT A3IA'] = 'MGHHHHHHMGGSTPWDSPSMAESFMRSFINGISSSGAFSGDQIGDMQDITGTMQASVEKMASTGRSSKSKLQAMNMAFASSMAEIAAAEAGGASMDSKTNAITDALRGAFLQTTGVSNEQFITEIRGLVSLIASNVNAGNSGRGQGGYGQGSGGNAAAIAAAIAAAIAAAGQGGQGGYGRQSQGAGSAAAAAAAAAAAAAAGSGQGGYGGQGQGGYGQSGSVTSGGYGYGTSAAAGAGVAAGSYAGAVNRLSSAEAASRVSSNIAAIASGGASALPSVISNIYSGVVASGVSSNEALIQALLELLSALVHVLSSASIGNVSSVGVDSTLNVVQDSVGQYVG'
    seqs_ds['Br_MaSp2_300'] = seq_nt + 'GGPGASAAVAVSSGPGGYGPGSPGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGSSAAVAVSSGPGGYGPGSQGGPSGPGSQGPSGPGGPGSSSAASGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGASAAVAVSSGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGASAAVAVSSGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGSSAAVAVSSGPGGYGPGSQGGPSGPGSQGPSG' + seq_ct
    seqs_ds['Br_MaSp2_400'] = seq_nt + 'GGPGASAAVAVSSGPGGYGPGSPGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGSSAAVAVSSGPGGYGPGSQGGPSGPGSQGPSGPGGPGSSSAASGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGASAAVAVSSGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGASAAVAVSSGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGSSAAVAVSSGPGGYGPGSQGGPSGPGSQGPSGPGGPGSSSAASGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGASAAVAVSSGPGGYGPGSQGPSGPSGPGGYGPGSQGGPS' + seq_ct
    seqs_ds['Br_MaSp2_long'] = seq_nt + 'GGPGASAAVAVSSGPGGYGPGSPGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGSSAAVAVSSGPGGYGPGSQGGPSGPGSQGPSGPGGPGSSSAASGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGASAAVAVSSGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGASAAVAVSSGPGGYGPGSQGP' + seq_ct
    seqs_ds['Br_MaSp2_short'] = seq_nt + 'SAASGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGASAAVAVSSGPGGYGPGSQGPSGPSGPGGYGPGSQGGPSGPGGYGPGSQGPNGPGGPGASAAVAVSSGPGGYGPGSQ' + seq_ct

    fasta_path = "../data/spider-silkome-database.v1.prot.fasta"

    seqs_silkome = {}

    for record in SeqIO.parse(fasta_path, "fasta"):
        seq_id = record.id          # e.g. "SeqID_123"
        sequence = str(record.seq)  # amino acid sequence
        seqs_silkome[seq_id] = sequence

    print(f"Loaded {len(seqs_silkome)} sequences")# Store sequences by ID

    k=100
    for ds_id, ds_seq in seqs_ds.items():
        matches = match_sequences(ds_id, ds_seq, seqs_silkome, top_k=k)
        df_matches = matches_to_dataframe(matches)
        df_matches.to_csv(f"../data/sequence_matches/top_{k}_sequence_matches_{ds_id}.csv", index=False)